hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java

```diff
@@ -1979,10 +1979,61 @@ LocatedBlocks getBlockLocations(String clientMachine, String srcArg,
         blockManager.getDatanodeManager().sortLocatedBlocks(
             clientMachine, lastBlockList);
       }
+
+      //Locations of LocatedBlocks have been sorted
+      //Measure distance to first location of all block to get a feeling of read locality within the cluster
+      for(LocatedBlock b : blocks.getLocatedBlocks()) {
+        measureDistanceToFirstLocation(clientMachine, b);
+      }
+      // lastBlock is not part of getLocatedBlocks(), so we need to check it individually
+      if (lastBlock != null) {
+        measureDistanceToFirstLocation(clientMachine, lastBlock);
+      }
     }
+
     return blocks;
   }
 
+  private void measureDistanceToFirstLocation(String clientMachine, LocatedBlock locatedBlock) {
+    DatanodeInfo[] locs = locatedBlock.getLocations();
+    if(locs.length > 0) {
+      DatanodeInfo closerLocation = locs[0];
+      NetworkTopology networkTopology = blockManager.getDatanodeManager().getNetworkTopology();
+
+      //mimic code in DatanodeManager.sortLocatedBlocks
+      //to be able to create a Node object representing possibly an offswitch client
+      //then we can use NetworkTopology to measure the distance
+      Node client = blockManager.getDatanodeManager().getDatanodeByHost(clientMachine);
+      if (client == null) {
+        List<String> hosts = new ArrayList<String> (1);
+        hosts.add(clientMachine);
+        List<String> resolvedHosts = blockManager.getDatanodeManager().resolveNetworkLocation(hosts);
+        if (resolvedHosts != null && !resolvedHosts.isEmpty()) {
+          String rName = resolvedHosts.get(0);
+          if (rName != null) {
+            client = new NodeBase(rName + NodeBase.PATH_SEPARATOR_STR +
+                    clientMachine);
+          }
+        } else {
+          LOG.error("Node Resolution failed. Please make sure that rack " +
+                  "awareness scripts are functional.");
+        }
+      }
+
+      //client could be null due to the code above, but getWeight method accepts null first parameter
+      int weight = networkTopology.getWeight(client, closerLocation);
+      //describe in method getWeight: 0 is local, 1 is same rack, 2 is off rack
+      ReadLocalityMetrics readLocalityMetrics = NameNode.getReadLocalityMetrics();
+      if (weight == 0) {
+        readLocalityMetrics.incrHostLocalReads();
+      } else if (weight == 1) {
+        readLocalityMetrics.incrRackLocalReads();
+      } else {
+        readLocalityMetrics.incrOffSwitchReads();
```

Comment:
s/offSwitch/offRack ?

Reply:
Generally in Hadoop code they use offswitch to talk about out of local/rack 
Ex https://github.com/criteo-forks/hadoop-common/blob/433a60a503760c819ddae60add928a9ae7dfeff3/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/capacity/LeafQueue.java#L1234